src/platform/cocoamain.mm

```diff
@@ -1218,6 +1218,182 @@ - (IBAction)preferences:(id)sender {
+    uint16_t version;
+    uint16_t client;
+    uint16_t command;
+    int16_t param;
+    int32_t value;
+    UInt64 time;
+    uint8_t report[8];
+    uint16_t buttons8;
+    int16_t axis[6];
+    uint16_t address;
+    uint32_t buttons;
+} ConnexionDeviceState, *ConnexionDeviceStatePtr;
+
+#pragma pack(pop)
+
+typedef void (*ConnexionAddedHandlerProc)(io_connect_t);
+typedef void (*ConnexionRemovedHandlerProc)(io_connect_t);
+typedef void (*ConnexionMessageHandlerProc)(io_connect_t, natural_t, void *);
+
+typedef OSErr (*InstallConnexionHandlersProc)(ConnexionMessageHandlerProc, ConnexionAddedHandlerProc, ConnexionRemovedHandlerProc);
+typedef void (*CleanupConnexionHandlersProc)(void);
+typedef UInt16 (*RegisterConnexionClientProc)(UInt32, UInt8 *, UInt16, UInt32);
+typedef void (*UnregisterConnexionClientProc)(UInt16);
+
+static BOOL connexionShiftIsDown = NO;
+static UInt16 connexionClient = 0;
+static UInt32 connexionSignature = 'SoSp';
+static UInt8 *connexionName = (UInt8 *)"SolveSpace";
+static CFBundleRef spaceBundle = NULL;
+static InstallConnexionHandlersProc installConnexionHandlers = NULL;
+static CleanupConnexionHandlersProc cleanupConnexionHandlers = NULL;
+static RegisterConnexionClientProc registerConnexionClient = NULL;
+static UnregisterConnexionClientProc unregisterConnexionClient = NULL;
+
+/*
+ * Space Mouse events are generated form another Thread and
+ * need to be sent to the main thread, so we have to use
+ * [NSObject performSelectorOnMainThread]. Unfortunately,
+ * we want to use this with a C++ and not an Objective-C object,
+ * so we need an Objective-C wrapper object to use it.
+ */
+@interface SpaceMouseObject : NSObject {
+    double data0, data1, data2;
+    double data3, data4, data5;
+}
+@property double data0, data1, data2;
+@property double data3, data4, data5;
+- (void)invoke;
+@end
+
+@implementation SpaceMouseObject
+@synthesize data0, data1, data2;
+@synthesize data3, data4, data5;
+- (void)invoke {
+    SolveSpace::SS.GW.SpaceNavigatorMoved(
+        data0, data1, data2,
+        data3, data4, data5,
+        (connexionShiftIsDown == YES) ? 1 : 0
+    );
+}
+@end
+
+static void connexionAdded(io_connect_t con) {}
+static void connexionRemoved(io_connect_t con) {}
+static void connexionMessage(io_connect_t con, natural_t type, void *arg) {
+    if (type != kConnexionMsgDeviceState) {
+        return;
+    }
+
+    ConnexionDeviceState *device = (ConnexionDeviceState *)arg;
+
+    SpaceMouseObject *space = [[SpaceMouseObject alloc] init];
+    space.data0 = (double)device->axis[0] * -0.25;
+    space.data1 = (double)device->axis[1] * -0.25;
+    space.data2 = (double)device->axis[2] * 0.25;
+    space.data3 = (double)device->axis[3] * -0.0005;
+    space.data4 = (double)device->axis[4] * -0.0005;
+    space.data5 = (double)device->axis[5] * -0.0005;
+    [space performSelectorOnMainThread:@selector(invoke)
+                            withObject:nil
+                         waitUntilDone:NO];
+}
```

Comment:
Ok--one last change. Let's use `dispatch_async(dispatch_get_main_queue(), ^(void){...})` instead of this object used once; `dispatch_async` seems to be the recommended API anyway.